architecture-docs/adr-003.md

```diff
@@ -0,0 +1,144 @@
+# ADR 003: Migrate to jellyfish merkle trie
+
+## Changelog
+* 31-03-2020: Initial Draft
+
+## Context
+
+Need a merkle trie library to store staking state, current one has several disadvantages:
+
+- Written and maintained by single person.
+- Custom data structure and algorithm.
+- Not enough documentation.
+
+## Decision
+
+### jellyfish merkle trie
+
+https://github.com/libra/libra/blob/master/storage/jellyfish-merkle/src/lib.rs
+
+#### Advantages
+
+- Review from Facebook security engineers
+- Small code base, well documented.
+- Well known data structure (simplified version of ethereum MPT).
+- Easy to integrate.
+- Optimized for pbft: Use u64 version number instead of hash.
+- License is apache 2.0.
+- Support inclusion proof and range proof.
+
+#### Disadvantages
+
+- Tied with a lot of custom Libra crates, need to extract an independent library.
+
+  Initial work has done here: https://github.com/yihuang/libra
+
+- No extension node compared to ethereum MPT, deeper tree in theory, a trade off made in favor of simplicity.
+
+### Integration
+
+- Maintain a custom fork of libra repository, make nessesary changes to keep jellyfish-merkle separated from other part
+  of the project.
+
+  Sync the code periodically manually with git.
+
+- Add a column `COL_TRIE_NODE` to store trie nodes in the same key-value store as other states.
+
+- Implement `jellyfish_merkle::TreeReader` for `GetKV`.
+
+  ```rust
+  struct KVReader<'a, S: GetKV>(&'a S);
+  impl<'a, S: GetKV> TreeReader for KVReader<'a, S> {
+      fn get_node_option(&self, node_key: &NodeKey) -> Result<Option<Node>> {
+          self.0
+              .get(&(COL_TRIE_NODE, node_key.encode()?))
+              .map(|bytes| Node::decode(&bytes))
+              .transpose()
+      }
+  
+      fn get_rightmost_leaf(&self) -> Result<Option<(NodeKey, LeafNode)>> {
+          unimplemented!("this feature is only used in merkle tree restore which we don't need yet");
+      }
+  }
+  ```
+
+- Implement `StakingGetter` with `JellyfishMerkleTree` against any `GetKV`,
+
+  Use `BlockHeight` as the version number.
+  
+  ```rust
+  pub struct StakingGetter<'a, S: GetKV> {
+      storage: &'a S,
+      block_height: BlockHeight,
+  }
+  
+  impl<'a, S: GetKV> Get for StakingGetter<'a, S> {
+      type Key = StakedStateAddress;
+      type Value = StakedState;
+      fn get(&self, key: &Self::Key) -> Option<Self::Value> {
```

Comment:
what about getting/returning the proof?

Reply:
> what about getting/returning the proof?

Just use `JellyfishMerkleTree` api directly then.